### PR TITLE
Fix Use-of-uninitialized-value in std::__1::basic_string<char, std::_…

### PR DESCRIPTION
**Pull request by sashashura**
…_1::char_traits<char>, std::__1::allocator<ch

https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=23922

If `IDnsResource::decodeName` doesn't null terminate the `result` at the calculated length it leads to uninitialized buffer read later in a string assignment.

### CODE REVIEW

**Comment by sashashura** [comment]
```
==489==WARNING: MemorySanitizer: use-of-uninitialized-value
    #1 0x6e8e93 in std::_Rb_tree<unsigned short, std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned short>, std::allocator<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::lower_bound(unsigned short const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_tree.h:1187:16
    #3 0x6dde52 in std::map<unsigned short, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<unsigned short>, std::allocator<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::operator[](unsigned short&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_map.h:509:17
    #4 0x6d7796 in pcpp::createPPPNextProtoToStringMap[abi:cxx11]() /tmp/cirrus-ci-build/Packet++/src/PPPoELayer.cpp:101:2
    #5 0x4328a9 in __cxx_global_var_init.130 /tmp/cirrus-ci-build/Packet++/src/PPPoELayer.cpp:236:62
    #6 0x432949 in _GLOBAL__sub_I_PPPoELayer.cpp /tmp/cirrus-ci-build/Packet++/./src/PPPoELayer.cpp
    #7 0x7d9b3c in __libc_csu_init (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x7d9b3c)
SUMMARY: MemorySanitizer: use-of-uninitialized-value /usr/bin/../lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_tree.h:1871:7 in std::_Rb_tree<unsigned short, std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned short>, std::allocator<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_lower_bound(std::_Rb_tree_node<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*, std::_Rb_tree_node_base*, unsigned short const&)
```
That's odd. It looks like a different bug reproducible with the same sample file. https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=23922 stack is 
```
  std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<ch
  std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<ch
  pcpp::IDnsResource::IDnsResource
```
where the `decodeName((const char*)getRawData(), decodedName);` is called.
However I have difficulties reproducing the new crash in my environment.
I propose not committing the reproducer file to keep tests green, but create a separate issue for the new crash.
What do you think?

File: Packet++/src/DnsResource.cpp [hunk 1]
```diff
@@ -56,7 +56,7 @@ size_t IDnsResource::decodeName(const char* encodedName, char* result, int itera
 		if ((wordLength & 0xc0) == 0xc0)
 		{
 			if (curOffsetInLayer + 2 > m_DnsLayer->m_DataLen || encodedNameLength > 255)
-				return encodedNameLength;
+				goto exit;
```
Comment (seladb) [inline Packet++/src/DnsResource.cpp:59]: `goto` is [not encouraged in c++](https://stackoverflow.com/questions/46586/goto-still-considered-harmful), I think we should try to avoid it...

**Review by seladb — CHANGES_REQUESTED** [review]
I'm not sure what are the other changes other than fixing `DnsResource.cpp`, but if they're needed I think we should move them to another PR...

**Comment by seladb** [comment]
> ```
> ==489==WARNING: MemorySanitizer: use-of-uninitialized-value
>     #1 0x6e8e93 in std::_Rb_tree<unsigned short, std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned short>, std::allocator<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::lower_bound(unsigned short const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_tree.h:1187:16
>     #3 0x6dde52 in std::map<unsigned short, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<unsigned short>, std::allocator<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::operator[](unsigned short&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_map.h:509:17
>     #4 0x6d7796 in pcpp::createPPPNextProtoToStringMap[abi:cxx11]() /tmp/cirrus-ci-build/Packet++/src/PPPoELayer.cpp:101:2
>     #5 0x4328a9 in __cxx_global_var_init.130 /tmp/cirrus-ci-build/Packet++/src/PPPoELayer.cpp:236:62
>     #6 0x432949 in _GLOBAL__sub_I_PPPoELayer.cpp /tmp/cirrus-ci-build/Packet++/./src/PPPoELayer.cpp
>     #7 0x7d9b3c in __libc_csu_init (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x7d9b3c)
> SUMMARY: MemorySanitizer: use-of-uninitialized-value /usr/bin/../lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_tree.h:1871:7 in std::_Rb_tree<unsigned short, std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned short>, std::allocator<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_lower_bound(std::_Rb_tree_node<std::pair<unsigned short const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*, std::_Rb_tree_node_base*, unsigned short const&)
> ```
> 
> That's odd. It looks like a different bug reproducible with the same sample file. https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=23922 stack is
> 
> ```
>   std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<ch
>   std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<ch
>   pcpp::IDnsResource::IDnsResource
> ```
> 
> where the `decodeName((const char*)getRawData(), decodedName);` is called. However I have difficulties reproducing the new crash in my environment. I propose not committing the reproducer file to keep tests green, but create a separate issue for the new crash. What do you think?

Yes, I think that'd be a good idea

**Comment by sashashura** [comment]
> I'm not sure what are the other changes other than fixing `DnsResource.cpp`, but if they're needed I think we should move them to another PR...

Do you mean the moving around the reproducer folders for specific sanitizers? If I add the reproducer file for this case it passes the memory sanitizer but makes the address sanitizer tests red. What if I'll not add the reproducer files in the PR at all, just the fix. Then will create a separate PR for the folder reorganization and a separate issue to investigate the crash in `createPPPNextProtoToStringMap`. OK?

File: Packet++/src/DnsResource.cpp (hunk 1 above)
Comment (sashashura) [reply Packet++/src/DnsResource.cpp:59]: Usually it is true and I am also very against goto. But in this case it would mean copy/paste of the following code here and goto cleanup is the only acceptable pattern imho:
```cpp
        if (resultPtr > result)
	{
		result[resultPtr - result - 1] = 0;
	}
	if (resultPtr - result < 256)
	{
		// add the last '\0' to encodedNameLength
		resultPtr[0] = 0;
		encodedNameLength++;
	}
	return encodedNameLength;
```
Which makes more code duplication and less readable code.
Let me know what you think.

Comment (seladb) [reply Packet++/src/DnsResource.cpp:59]: We can move these lines to a new private method and then call it twice 🤔

**Comment by seladb** [comment]
> > I'm not sure what are the other changes other than fixing `DnsResource.cpp`, but if they're needed I think we should move them to another PR...
> 
> Do you mean the moving around the reproducer folders for specific sanitizers? If I add the reproducer file for this case it passes the memory sanitizer but makes the address sanitizer tests red. What if I'll not add the reproducer files in the PR at all, just the fix. Then will create a separate PR for the folder reorganization and a separate issue to investigate the crash in `createPPPNextProtoToStringMap`. OK?

Would you consider fixing the other bug in this PR so it passes both memory, address an undefined santizers?

File: Packet++/src/DnsResource.cpp (hunk 1 above)
Comment (sashashura) [reply Packet++/src/DnsResource.cpp:59]: Ok, will do.

**Comment by sashashura** [comment]
> Would you consider fixing the other bug in this PR so it passes both memory, address an undefined santizers?

It was mistake on my side. I was running the address sanitizer, but forgot to change compilation flags.

**Comment by sashashura** [comment]
Everything is done:
1. All tests fail with memory sanitizer because of the same `pcpp::createPPPNextProtoToStringMap` mentioned above, so I have removed the memory sanitizers run which I added for this `clusterfuzz-testcase-minimized-FuzzTarget-5739357070950400` case.
2. The `clusterfuzz-testcase-minimized-FuzzTarget-5739357070950400` fails address sanitizer. Looks like out of memory in LibFuzzer itself because of CI restrictions. I cannot reproduce it locally :(. I suggest pushing the fix only without the test case.
```
Running sample clusterfuzz-testcase-minimized-FuzzTarget-5739357070950400...INFO: Seed: 2836457384
INFO: Loaded 1 modules   (14266 inline 8-bit counters): 14266 [0x9ea720, 0x9ededa), 
INFO: Loaded 1 PC tables (14266 PCs): 14266 [0x9edee0,0xa25a80), 
../Bin/FuzzTarget: Running 1 inputs 1 time(s) each.
Running: regression_samples/clusterfuzz-testcase-minimized-FuzzTarget-5739357070950400
==550== ERROR: libFuzzer: out-of-memory (malloc(4294967072))
   To change the out-of-memory limit use -rss_limit_mb=<N>
    #0 0x4f3891 in __sanitizer_print_stack_trace (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x4f3891)
    #1 0x47292b in fuzzer::PrintStackTrace() (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x47292b)
    #2 0x456905 in fuzzer::Fuzzer::HandleMalloc(unsigned long) (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x456905)
    #3 0x45680a in fuzzer::MallocHook(void const volatile*, unsigned long) (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x45680a)
    #4 0x4f9bf7 in __sanitizer::RunMallocHooks(void const*, unsigned long) (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x4f9bf7)
    #5 0x474db5 in __asan::Allocator::Allocate(unsigned long, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType, bool) (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x474db5)
    #6 0x474583 in __asan::asan_malloc(unsigned long, __sanitizer::BufferedStackTrace*) (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x474583)
    #7 0x4eb03b in malloc (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x4eb03b)
    #8 0x7f2634e2290f  (/usr/lib/x86_64-linux-gnu/libpcap.so.0.8+0x1f90f)
    #9 0x7f2634e21ecc in pcap_fopen_offline_with_tstamp_precision (/usr/lib/x86_64-linux-gnu/libpcap.so.0.8+0x1eecc)
    #10 0x7f2634e220cd in pcap_open_offline_with_tstamp_precision (/usr/lib/x86_64-linux-gnu/libpcap.so.0.8+0x1f0cd)
    #11 0x523363 in pcpp::PcapFileReaderDevice::open() /tmp/cirrus-ci-build/Pcap++/src/PcapFileDevice.cpp:254:21
    #12 0x51d50a in LLVMFuzzerTestOneInput /tmp/cirrus-ci-build/Tests/Fuzzers/FuzzTarget.cpp:43:14
    #13 0x458c61 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x458c61)
    #14 0x441157 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x441157)
    #15 0x447131 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x447131)
    #16 0x473192 in main (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x473192)
    #17 0x7f2633e52c86 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21c86)
    #18 0x43b549 in _start (/tmp/cirrus-ci-build/Tests/Fuzzers/Bin/FuzzTarget+0x43b549)
SUMMARY: libFuzzer: out-of-memory
```

File: Tests/Fuzzers/RegressionTests/run_tests.sh [hunk 2]
```diff
@@ -10,7 +10,11 @@ NC='\033[0m'
 
 for sample in $(ls ${SAMPLES}); do
 	echo -n "Running sample $sample..."
-	$BINARY $SAMPLES/$sample &> /dev/null && echo -e "${GREEN}[OK]${NC}" || { FAILED=True && echo -e "${RED}[FAIL]${NC}"; }
+	if [ -z "$1" ]; then
+		$BINARY $SAMPLES/$sample &> /dev/null && echo -e "${GREEN}[OK]${NC}" || { FAILED=True && echo -e "${RED}[FAIL]${NC}"; }
+	else
+		$BINARY $SAMPLES/$sample && echo -e "${GREEN}[OK]${NC}" || { FAILED=True && echo -e "${RED}[FAIL]${NC}"; }
+	fi
```
Comment (seladb) [inline Tests/Fuzzers/RegressionTests/run_tests.sh:17]: Is this change needed in this PR? 🤔

Comment (sashashura) [reply Tests/Fuzzers/RegressionTests/run_tests.sh:17]: That was useful to see the output when tests were failing, but I can remove it if you want.

Comment (seladb) [reply Tests/Fuzzers/RegressionTests/run_tests.sh:17]: I think we can remove it, let me know if you think otherwise

Comment (sashashura) [reply Tests/Fuzzers/RegressionTests/run_tests.sh]: I have removed it. Anyway I need to update `.cirrus.yml` if I want to debug the output.

**Review by seladb — APPROVED** [review]
@sashashura thanks again for your great contribution to this project! 🙏 ❤️
